crates/amaru/src/consensus/chain_selection.rs

```diff
@@ -0,0 +1,421 @@
+            None => self.anchor.clone(),
+        }
+    }
+
+    fn extend_with(&mut self, header: &H) -> FragmentExtension {
+        if header.parent().unwrap() == self.tip().hash() {
+            self.headers.push(header.clone());
+            FragmentExtension::Extend
+        } else {
+            FragmentExtension::Ignore
+        }
+    }
+}
+
+/// Current state of chain selection process
+///
+/// Chain selection is parameterised by the header type `H`, in
+/// order to better decouple the internals of what's a header from
+/// the selection logic
+pub struct ChainSelector<H: Header> {
+    tip: H,
+    peers_chains: HashMap<Peer, Fragment<H>>,
+}
+
+/// The outcome of the chain selection process.
+#[derive(Debug, PartialEq)]
+pub enum ChainSelection<H: Header> {
+    /// The current best chain has been extended with a (single) new header.
+    NewTip(H),
+
+    /// The current best chain has been rolled back to the given hash.
+    RollbackTo(Hash<32>),
+
+    /// The current best chain is unchanged.
+    NoChange,
+
+    /// The current best chain has switched to given fork starting at
+    /// given hash.
+    SwitchToFork(Point, Vec<H>),
+}
+
+/// Builder pattern for `ChainSelector`.
+///
+/// Allows incrementally adding information to build a
+/// fully functional `ChainSelector`.
+pub struct ChainSelectorBuilder<H: Header> {
+    tip: Option<H>,
+    peers: Vec<Peer>,
+}
+
+impl<H: Header + Clone> ChainSelectorBuilder<H> {
+    pub fn new() -> ChainSelectorBuilder<H> {
+        ChainSelectorBuilder {
+            tip: None,
+            peers: Vec::new(),
+        }
+    }
+
+    pub fn set_tip(&mut self, new_tip: &H) -> &mut Self {
+        self.tip = Some(new_tip.clone());
+        self
+    }
+
+    pub fn add_peer(&mut self, peer: &Peer) -> &mut Self {
+        self.peers.push(peer.clone());
+        self
+    }
+
+    pub fn build(&self) -> ChainSelector<H> {
+        ChainSelector {
+            tip: self
+                .tip
+                .clone()
+                .expect("cannot build a chain selector without a tip"),
+            peers_chains: self
+                .peers
+                .iter()
+                .map(|peer| {
+                    (
+                        peer.clone(),
+                        Fragment::start_from(self.tip.as_ref().unwrap()),
+                    )
+                })
+                .collect(),
+        }
+    }
+}
+
+impl<H: Header + Clone> Default for ChainSelectorBuilder<H> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<H> ChainSelector<H>
+where
+    H: Header + Clone + Debug + PartialEq,
+{
+    /// Roll forward the chain with a new header from given peer.
+    ///
+    /// The function returns the result of the chain selection process, which might lead
+    /// to a new tip, a switch to a fork, no change, or some change in status for the peer.
+    #[instrument(skip(self, header), fields(slot = header.slot(), hdr = header.hash().to_string()))]
```

Comment:
We might want to agree on this syntax for instrumentation.

Reply:
Could you elaborate?

Reply:
I mean use `instrument` instead of manually creating spans as a standard practise. So maybe a candidate for an AR?

Reply:
sure, we should strive to remove the o11y clutter from the code.